mk/Common.mk

```diff
@@ -231,123 +261,157 @@ include $(CFG_FLAGS_MAKE)
-CV_SW_MARCH         ?= $(call RESOLVE_FLAG2,$(CFG_CV_SW_MARCH),rv32imc)
-
-GNU_YES          = $(call IS_YES,$(GNU))
-PULP_YES         = $(call IS_YES,$(PULP))
-COREV_YES        = $(call IS_YES,$(COREV))
-LLVM_YES         = $(call IS_YES,$(LLVM))
-
-ifeq ($(shell $(CORE_V_VERIF)/mk/toolchain_check.sh $(GNU_YES) $(PULP_YES) $(COREV_YES) $(LLVM_YES)),1)
-$(error Multiple toolchains are enabled: GNU=${GNU_YES} PULP=${PULP_YES} COREV=${COREV_YES} LLVM=${LLVM_YES})
-endif
-
-RISCV             = $(CV_SW_TOOLCHAIN)
-RISCV_PREFIX      = riscv32-$(CV_SW_VENDOR)-elf-
-RISCV_EXE_PREFIX  = $(RISCV)/bin/$(RISCV_PREFIX)
-RISCV_CC          = gcc
-RISCV_MARCH       = $(call RESOLVE_FLAG2,$(TEST_RISCV_MARCH),$(CV_SW_MARCH))
-RISCV_CFLAGS      = $(TEST_RISCV_CFLAGS)
-
-ifeq ($(GNU_YES),YES)
-ifeq ($(call IS_YES,$(TEST_GNU_NOT_SUPPORTED)),YES)
-$(error test [$(TEST)] does not support the GNU toolchain)
-endif
-RISCV            = $(GNU_SW_TOOLCHAIN)
-RISCV_PREFIX     = riscv32-$(GNU_VENDOR)-elf-
+# The so-called "COREV Environment Variables".
+#
+# First, check that the COREV environment variables have been defined.
+ifndef CV_SW_TOOLCHAIN
+$(error Must define CV_SW_TOOLCHAIN)
+endif
+ifndef CV_SW_PREFIX
+$(error Must define CV_SW_PREFIX)
+endif
+ifndef CV_SW_MARCH
+$(error Must define CV_SW_MARCH)
+endif
+#ifndef CV_SW_CFLAGS
+#$(warning CV_SW_CFLAGS is optional)
+#endif
+
+# Use these to set defaults for toolchain vars. A few typical examples...
+#RISCV            = /opt/riscv
+#RISCV_PREFIX     = riscv32-unknown-elf-
+#RISCV_EXE_PREFIX = $(RISCV)/bin/$(RISCV_PREFIX)
+#RISCV_CFLAGS     =
+RISCV            = $(CV_SW_TOOLCHAIN)
+RISCV_PREFIX     = $(CV_SW_PREFIX)
 RISCV_EXE_PREFIX = $(RISCV)/bin/$(RISCV_PREFIX)
-RISCV_CC         = $(GNU_CC)
-RISCV_MARCH      = $(call RESOLVE_FLAG2,$(TEST_GNU_MARCH),$(GNU_MARCH))
-RISCV_CFLAGS     = $(call RESOLVE_FLAG2,$(TEST_GNU_CFLAGS),$(GNU_CFLAGS))
-endif
 
-ifeq ($(COREV_YES),YES)
-ifeq ($(call IS_YES,$(TEST_COREV_NOT_SUPPORTED)),YES)
-$(error test [$(TEST)] does not support the COREV toolchain)
-endif
-RISCV            = $(COREV_SW_TOOLCHAIN)
-RISCV_PREFIX     = riscv32-$(COREV_VENDOR)-elf-
-RISCV_EXE_PREFIX = $(RISCV)/bin/$(RISCV_PREFIX)
-RISCV_CC         = $(COREV_CC)
-RISCV_MARCH      = $(call RESOLVE_FLAG2,$(TEST_COREV_MARCH),$(COREV_MARCH))
-RISCV_CFLAGS     = $(call RESOLVE_FLAG2,$(TEST_COREV_CFLAGS),$(COREV_CFLAGS))
-endif
+RISCV_CC         = gcc
+RISCV_MARCH      = rv32imc
 
-ifeq ($(PULP_YES),YES)
-ifeq ($(call IS_YES,$(TEST_PULP_NOT_SUPPORTED)),YES)
-$(error test [$(TEST)] does not support the PULP toolchain)
-endif
-RISCV            = $(PULP_SW_TOOLCHAIN)
-RISCV_PREFIX     = riscv32-$(PULP_VENDOR)-elf-
-RISCV_EXE_PREFIX = $(RISCV)/bin/$(RISCV_PREFIX)
-RISCV_CC         = $(PULP_CC)
-RISCV_MARCH      = $(call RESOLVE_FLAG2,$(TEST_PULP_MARCH),$(PULP_MARCH))
-RISCV_CFLAGS     = $(call RESOLVE_FLAG2,$(TEST_PULP_CFLAGS),$(PULP_CFLAGS))
-endif
+RISCV_MARCH      = $(call RESOLVE_FLAG2,$(TEST_RISCV_MARCH),$(COREV_SW_MARCH))
```

Comment:
I think this "COREV_SW_MARCH" should be "CV_SW_MARCH"?
Same with "COREV_SW_CFLAGS" below.

Reply:
While we are changing these fundamental parts of the makefile, i would like to propose another change as well, and that is to add a function that resolves more than two alternatives with a prioritized order (e.g. CFG > TEST > ENV > default)

Reply:
I agree that we should have a prioritized order. I just don't know what it should be.

This is what I was trying to do with the code in:
https://github.com/openhwgroup/core-v-verif/pull/1092#discussion_r778016545
```
RISCV_MARCH = rv32imc
RISCV_MARCH = $(call RESOLVE_FLAG2,$(CFG_COREV_MARCH),$(RISCV_MARCH))
RISCV_MARCH = $(call RESOLVE_FLAG2,$(TEST_RISCV_MARCH),$(RISCV_MARCH))
RISCV_MARCH = $(call RESOLVE_FLAG2,$(CV_SW_MARCH),$(RISCV_MARCH))
```
(If that code works then maybe a fixed-arity function isn't needed?)

The order is also not clear to me. Is CFG over TEST?
I thought maybe: ENV > TEST > CFG.
The CFGs serve as "default", and we shouldn't need makefile-given defaults since we already have "default.yaml" (which should be mandatory for any core directory in core-v-verif?).
While TESTs usually don't specify any env variables at all, some tests might need to have particular settings that have to be met regardless of the config in use.
Then ENV can be user-given whenever we need to turn the knobs to debug something specific, or to match our particular computer setup.

Reply:
I tend to think that TEST should override other settings, as that is required by that specific test, while the general setting should be in CFG. I am not a big fan of having to set these things in the environment, that makes the entire setup less portable. If there is a common place to create these overrides (my suggestion: CFG), then we would have less of a dependency on system-specific variables. That would also lead to having a common place to configure the entire test-environment. 

The way things are now, some things have to be set by env vars, others in the CFG yamls and finally some things in the test yamls. When I first started out, it was not at all clear to me where certain things were configured, so the more we can confine the configuration to certain spaces, the easier it will be to use. 

Only very system specific things such as compiler locations/naming and simulator location/naming really warrants an environment-setting. Things that are self-contained within core-v-verif should, in my opinion, be located in either CFG or TEST yaml files, depending on scope of the parameter.

Reply:
@silabs-robin said:
> I thought maybe: ENV > TEST > CFG

Yes, that is the correct order.

Reply:
> @silabs-robin said:
> 
> > I thought maybe: ENV > TEST > CFG
> 
> Yes, that is the correct order.

Will this always be the case? I generally agree, but with slight reservations. For compiler paths and the like I agree, but for things like march we may want the test itself to be able to override a particular system-wide configuration, because we lose fine grained control.

Reply:
> for things like march we may want the test itself to be able to override a particular system-wide configuration

I agree that tests should be able to make test-specific overrides for special configs that it may need.
With "ENV > TEST > CFG" a test will be able to override the "system-wide" CFG, at least. And usually the ENV has simply not defined most of these variables?
(E.g. a CFG _should_ set march, a TEST _may_ override it, and the user-specified ENV usually never sets it.)